cdc/processor/sinkmanager/mem_quota.go

```diff
@@ -0,0 +1,149 @@
+// Copyright 2022 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package sinkmanager
+
+import (
+	"sort"
+	"sync"
+	"sync/atomic"
+
+	"github.com/pingcap/log"
+	"github.com/pingcap/tiflow/cdc/model"
+	cerrors "github.com/pingcap/tiflow/pkg/errors"
+	"go.uber.org/zap"
+)
+
+type memConsumeRecord struct {
+	resolvedTs model.ResolvedTs
+	size       uint64
+}
+
+type memQuota struct {
+	changefeedID model.ChangeFeedID
+	// totalBytes is the total memory quota for one changefeed.
+	totalBytes uint64
+
+	// mu protects the following fields.
+	mu sync.Mutex
+	// usedBytes is the memory usage of one changefeed.
+	usedBytes uint64
+	// tableMemory is the memory usage of each table.
+	tableMemory map[model.TableID][]*memConsumeRecord
+	// isClosed is used to indicate whether the mem quota is closed.
+	isClosed atomic.Bool
+	// blockAcquireCond is used to notify the blocked acquire.
+	blockAcquireCond *sync.Cond
+}
+
+func newMemQuota(changefeedID model.ChangeFeedID, totalBytes uint64) *memQuota {
+	m := &memQuota{
+		changefeedID: changefeedID,
+		totalBytes:   totalBytes,
+		usedBytes:    0,
+		tableMemory:  make(map[model.TableID][]*memConsumeRecord),
+	}
+	m.blockAcquireCond = sync.NewCond(&m.mu)
+
+	return m
+}
+
+func (m *memQuota) tryAcquire(nBytes uint64) bool {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	if m.usedBytes+nBytes > m.totalBytes {
+		return false
+	}
+	m.usedBytes += nBytes
+	return true
+}
+
+func (m *memQuota) forceAcquire(nBytes uint64) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	m.usedBytes += nBytes
+}
+
+func (m *memQuota) blockAcquire(nBytes uint64) error {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	for {
+		if m.isClosed.Load() {
+			return cerrors.ErrFlowControllerAborted.GenWithStackByArgs()
+		}
+
+		if m.usedBytes+nBytes <= m.totalBytes {
+			m.usedBytes += nBytes
+			return nil
+		}
+		m.blockAcquireCond.Wait()
+	}
+}
+
+func (m *memQuota) refund(nBytes uint64) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	m.usedBytes -= nBytes
```

Comment:
Should call `m.blockAcquireCond.Signal()`?

Reply:
Good catch!